pinot-broker/src/main/java/com/linkedin/pinot/broker/requesthandler/RangeMergeOptimizer.java

```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.broker.requesthandler;
+
+import com.linkedin.pinot.common.request.FilterOperator;
+import com.linkedin.pinot.common.utils.request.FilterQueryTree;
+import com.linkedin.pinot.core.common.predicate.RangePredicate;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.jar.Pack200;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+
+/**
+ * FilterOptimizer to merge intersecting range predicates:
+ * <ul>
+ *   <li> Given a filter query tree with range predicates on the same column, this optimizer merges the
+ *   predicates joined by AND by taking their intersection. </li>
+ *   <li> Pulls up merged predicates in the absence of other predicates on other columns. </li>
+ *   <li> Currently implemented to work for time column only. This is because the broker currently
+ *   does not know the data type for any columns, except for time column. </li>
+ * </ul>
+ */
+public class RangeMergeOptimizer extends FilterQueryTreeOptimizer {
+  private static final String DUMMY_STRING = "__dummy_string__";
+
+  @Override
+  public FilterQueryTree optimize(FilterQueryOptimizerRequest request) {
+    return optimizeRanges(request.getFilterQueryTree(), request.getTimeColumn());
+  }
+
+  /**
+   * Recursive method that performs the actual optimization of merging range predicates.
+   *
+   * @param current Current node being visited in the DFS of the filter query tree.
+   * @param timeColumn Name of time column
+   * @return Returns the optimized filter query tree
+   */
+  @Nonnull
+  private static FilterQueryTree optimizeRanges(@Nonnull FilterQueryTree current, @Nullable String timeColumn) {
+    if (timeColumn == null) {
+      return current;
+    }
+
+    List<FilterQueryTree> children = current.getChildren();
+    if (children == null || children.isEmpty()) {
+      return current;
+    }
+
+    // For OR, we optimize all its children, but do not propagate up.
+    if (current.getOperator() == FilterOperator.OR) {
+      int length = children.size();
+      for (int i = 0; i < length; i++) {
+        children.set(i, optimizeRanges(children.get(i), timeColumn));
+      }
+      return current;
+    }
+
+    List<FilterQueryTree> newChildren = new ArrayList<>();
```

Comment:
Can you add a comment here that says that after this point, since we have children, we're guaranteed to be an AND node (as well as an assert or if check, so that if we implement something like XOR that logic won't break).